File: src/main/java/de/tectoast/emolga/utils/YTDataLoader.java
package de.tectoast.emolga.utils;

import com.google.api.services.youtube.model.*;
import com.sedmelluq.discord.lavaplayer.track.AudioPlaylist;
import com.sedmelluq.discord.lavaplayer.track.AudioTrack;
import de.tectoast.emolga.utils.music.GuildMusicManager;
import net.dv8tion.jda.api.EmbedBuilder;
import net.dv8tion.jda.api.entities.Member;
import net.dv8tion.jda.api.entities.MessageEmbed;

import java.awt.*;

import static de.tectoast.emolga.commands.Command.formatToTime;

public class YTDataLoader {
    private final String channel;
    private final String thumbnail;
    private final String url;

    private YTDataLoader(Video video) {
        VideoSnippet sn = video.getSnippet();
        this.channel = sn.getChannelTitle();
        this.thumbnail = sn.getThumbnails().getMedium().getUrl();
        this.url = "https://www.youtube.com/watch?v=" + video.getId();
    }

    private YTDataLoader(SearchResult result) {
        SearchResultSnippet sn = result.getSnippet();
        this.channel = sn.getChannelTitle();
        this.thumbnail = sn.getThumbnails().getMedium().getUrl();
        this.url = "https://www.youtube.com/watch?v=" + result.getId().getVideoId();
    }

    private YTDataLoader(Playlist playlist) {
        PlaylistSnippet sn = playlist.getSnippet();
        this.channel = sn.getChannelTitle();
        this.thumbnail = sn.getThumbnails().getMedium().getUrl();
        this.url = "https://www.youtube.com/playlist?list=" + playlist.getId();
    }

    private static YTDataLoader fromURL(String url) {
        if (url.startsWith("https://www.youtube.com/playlist?list="))
            return new YTDataLoader(Google.getPlaylistByURL(url));
        if (url.startsWith("https://www.youtube.com/watch?v=") || url.startsWith("https://youtu.be/"))
            return new YTDataLoader(Google.getVidByURL(url));
        return null;
    }

    public static YTDataLoader create(String track) {
        if (track.startsWith("https://www.youtube.com/") || track.startsWith("https://youtu.be/")) {
            return fromURL(track);
        } else {
            return new YTDataLoader(Google.getVidByQuery(track));
        }
    }

    public String getChannel() {
        return channel;
    }

    public String getThumbnail() {
        return thumbnail;
    }

    public String getUrl() {
        return url;
    }

    public MessageEmbed buildEmbed(AudioTrack track, Member mem, GuildMusicManager musicManager) {
        long duration = musicManager.scheduler.queue.stream().mapToLong(AudioTrack::getDuration).sum() / 1000;
        if (musicManager.player.getPlayingTrack() != null) {
            duration += (musicManager.player.getPlayingTrack().getDuration() - musicManager.player.getPlayingTrack().getPosition());
        }
        return new EmbedBuilder()
                .setTitle(track.getInfo().title, url)
                .setColor(Color.CYAN)
                .setAuthor("Added to queue", null, mem.getUser().getEffectiveAvatarUrl())
                .addField("Channel", getChannel(), true)
                .addField("Song Duration", formatToTime(track.getDuration()), true)
                .addField("Estimated time until playing", formatToTime(duration), true)
                .addField("Position in queue", (musicManager.scheduler.queue.size() > 0 ? musicManager.scheduler.queue.size() + 1 : 0) + "", false)
                .setThumbnail(getThumbnail()).build();
    }

    public MessageEmbed buildEmbed(AudioPlaylist playlist, Member mem, GuildMusicManager musicManager) {
        long duration = musicManager.scheduler.queue.stream().mapToLong(AudioTrack::getDuration).sum() / 1000;
        if (musicManager.player.getPlayingTrack() != null) {
            duration += (musicManager.player.getPlayingTrack().getDuration() - musicManager.player.getPlayingTrack().getPosition());
        }
        return new EmbedBuilder()
                .setTitle(playlist.getName(), url)
                .setColor(Color.CYAN)
                .setAuthor("Added to queue", null, mem.getUser().getEffectiveAvatarUrl())
                .addField("Channel", getChannel(), true)
                .addField("Estimated time until playing", formatToTime(duration), true)
                .addField("Position in queue", (musicManager.scheduler.queue.size() > 0 ? musicManager.scheduler.queue.size() + 1 : 0) + "", true)
                .addField("Queued tracks", String.valueOf(playlist.getTracks().size()), true)
                .setThumbnail(getThumbnail()).build();
    }
}

File: src/main/kotlin/de/tectoast/emolga/utils/YTDataLoader.kt
package de.tectoast.emolga.utils

import com.google.api.services.youtube.model.Playlist
import com.google.api.services.youtube.model.SearchResult
import com.google.api.services.youtube.model.Video
import com.sedmelluq.discord.lavaplayer.track.AudioPlaylist
import com.sedmelluq.discord.lavaplayer.track.AudioTrack
import de.tectoast.emolga.commands.Command.Companion.formatToTime
import de.tectoast.emolga.utils.Google.getPlaylistByURL
import de.tectoast.emolga.utils.Google.getVidByQuery
import de.tectoast.emolga.utils.Google.getVidByURL
import de.tectoast.emolga.utils.music.GuildMusicManager
import net.dv8tion.jda.api.EmbedBuilder
import net.dv8tion.jda.api.entities.Member
import net.dv8tion.jda.api.entities.MessageEmbed
import java.awt.Color

class YTDataLoader {
    private val channel: String
    private val thumbnail: String
    val url: String

    private constructor(video: Video?) {
        val sn = video!!.snippet
        channel = sn.channelTitle
        thumbnail = sn.thumbnails.medium.url
        url = "https://www.youtube.com/watch?v=" + video.id
    }

    private constructor(result: SearchResult?) {
        val sn = result!!.snippet
        channel = sn.channelTitle
        thumbnail = sn.thumbnails.medium.url
        url = "https://www.youtube.com/watch?v=" + result.id.videoId
    }

    private constructor(playlist: Playlist?) {
        val sn = playlist!!.snippet
        channel = sn.channelTitle
        thumbnail = sn.thumbnails.medium.url
        url = "https://www.youtube.com/playlist?list=" + playlist.id
    }

    fun buildEmbed(track: AudioTrack, mem: Member, musicManager: GuildMusicManager): MessageEmbed {
        var duration = musicManager.scheduler.queue.stream().mapToLong { obj: AudioTrack -> obj.duration }.sum() / 1000
        if (musicManager.player.playingTrack != null) {
            duration += musicManager.player.playingTrack.duration - musicManager.player.playingTrack.position
        }
        return EmbedBuilder()
            .setTitle(track.info.title, url)
            .setColor(Color.CYAN)
            .setAuthor("Added to queue", null, mem.user.effectiveAvatarUrl)
            .addField("Channel", channel, true)
            .addField("Song Duration", formatToTime(track.duration), true)
            .addField("Estimated time until playing", formatToTime(duration), true)
            .addField(
                "Position in queue",
                (if (musicManager.scheduler.queue.size > 0) musicManager.scheduler.queue.size + 1 else 0).toString(),
                false
            )
            .setThumbnail(thumbnail).build()
    }

    fun buildEmbed(playlist: AudioPlaylist, mem: Member, musicManager: GuildMusicManager): MessageEmbed {
        var duration = musicManager.scheduler.queue.stream().mapToLong { obj: AudioTrack -> obj.duration }.sum() / 1000
        if (musicManager.player.playingTrack != null) {
            duration += musicManager.player.playingTrack.duration - musicManager.player.playingTrack.position
        }
        return EmbedBuilder()
            .setTitle(playlist.name, url)
            .setColor(Color.CYAN)
            .setAuthor("Added to queue", null, mem.user.effectiveAvatarUrl)
            .addField("Channel", channel, true)
            .addField("Estimated time until playing", formatToTime(duration), true)
            .addField(
                "Position in queue",
                (if (musicManager.scheduler.queue.size > 0) musicManager.scheduler.queue.size + 1 else 0).toString(),
                true
            )
            .addField("Queued tracks", playlist.tracks.size.toString(), true)
            .setThumbnail(thumbnail).build()
    }

    companion object {
        private fun fromURL(url: String): YTDataLoader? {
            if (url.startsWith("https://www.youtube.com/playlist?list=")) return YTDataLoader(getPlaylistByURL(url))
            return if (url.startsWith("https://www.youtube.com/watch?v=") || url.startsWith("https://youtu.be/")) YTDataLoader(
                getVidByURL(url)
            ) else null
        }

        fun create(track: String): YTDataLoader? {
            return if (track.startsWith("https://www.youtube.com/") || track.startsWith("https://youtu.be/")) {
                fromURL(track)
            } else {
                YTDataLoader(getVidByQuery(track))
            }
        }
    }
}